File: Python/graph_algorithms/07_is_bipartite.py
import sys
from collections import defaultdict, deque


def convert_to_intlist(line):
    return [int(elem) for elem in line.split()]


def read_edges(reader, nr_edges):
    edges = []
    for _ in range(nr_edges):
        edge = tuple(convert_to_intlist(next(reader)))
        edges.append(edge)
    return edges


class Graph(object):
    def __init__(self, edge_list, nr_nodes):
        self._adjacency_list = self._convert_edgelist_to_adjacency_list(edge_list)
        self._nr_nodes = nr_nodes
        self._coloring = [-1] * self._nr_nodes

    def _convert_edgelist_to_adjacency_list(self, edge_list):
        adjacency_list = defaultdict(list)
        for node_from, node_to in edge_list:
            adjacency_list[node_from].append(node_to)
            adjacency_list[node_to].append(node_from)
        return dict(adjacency_list)

    def _is_visited(self, node):
        return self._coloring[node-1] != -1

    def _get_consistently_colored_component(self, start_node):
        self._coloring[start_node-1] = 0
        queue = deque([start_node])
        component = {start_node}
        while queue:
            node = queue.popleft()
            node_color = self._coloring[node-1]
            neighbours = self._adjacency_list.get(node, [])
            for neighbour in neighbours:
                if self._is_visited(neighbour):
                    if node_color == self._coloring[neighbour-1]:
                        return
                else:
                    queue.append(neighbour)
                    self._coloring[neighbour-1] = 1 - node_color
                    component.add(neighbour)
        return component

    def breadth_first_search(self):
        unvisited_nodes = set(range(1, self._nr_nodes+1))
        while unvisited_nodes:
            start_node = unvisited_nodes.pop()
            bipartite_component = self._get_consistently_colored_component(start_node)
            if bipartite_component is None:
                return 0
            unvisited_nodes.difference_update(bipartite_component)
        return 1


def is_graph_bipartite(edge_list, nr_nodes):
    graph = Graph(edge_list, nr_nodes)
    return graph.breadth_first_search()


def main():
    reader = sys.stdin
    nr_nodes, nr_edges = convert_to_intlist(next(reader))
    edge_list = read_edges(reader, nr_edges)
    result = is_graph_bipartite(edge_list, nr_nodes)
    print(result)


if __name__ == "__main__":
    main()

File: Rust/graph_algorithms/src/bin/07_is_bipartite.rs
use std::collections::{HashMap, HashSet, VecDeque};
use std::io;

#[derive(Clone, Eq, PartialEq)]
enum NodeColor {
    Red,
    Blue,
}

impl NodeColor {
    fn opposite(&self) -> Self {
        match self {
            NodeColor::Blue => NodeColor::Red,
            NodeColor::Red => NodeColor::Blue,
        }
    }
}

struct Edge {
    start: i32,
    end: i32,
}

struct Graph {
    nr_nodes: i32,
    adjacency_list: HashMap<i32, HashSet<i32>>
}

impl Graph {
    fn from_edge_list(nr_nodes: i32, edges: &[Edge]) -> Self {
        let mut adjacency_list = HashMap::new();
        for &Edge { start, end } in edges {
            adjacency_list.entry(start).or_insert_with(HashSet::new).insert(end);
            adjacency_list.entry(end).or_insert_with(HashSet::new).insert(start);
        }

        Graph { nr_nodes, adjacency_list }
    }
}

fn convert_to_int_vector(line: &str) -> Vec<i32> {
    line
        .split_whitespace()
        .map(|x| x.parse().unwrap())
        .collect()
}

fn read_pair() -> (i32, i32) {
    let mut line = String::new();
    io::stdin().read_line(&mut line).unwrap();

    let xs: Vec<i32> = convert_to_int_vector(&line);
    let [a, b] = xs.as_slice() else {
        panic!("Malformed input: {}", line);
    };
    (*a, *b)
}

fn read_edges(nr_edges: usize) -> Vec<Edge> {
    let mut edges = Vec::with_capacity(nr_edges);
    for _ in 0..nr_edges {
        let (start, end) = read_pair();
        edges.push(Edge { start, end });
    }

    edges
}

fn get_consistently_colored_component(graph: &Graph, start_node: i32) -> Option<HashSet<i32>> {
    let mut coloring = HashMap::from([(start_node, NodeColor::Red)]);
    let mut component = HashSet::from([start_node]);
    let mut queue = VecDeque::from([start_node]);
    while let Some(node) = queue.pop_front() {
        let node_color = coloring.get(&node).unwrap().clone();
        if let Some(neighbors) = graph.adjacency_list.get(&node) {
            for &neighbor in neighbors {
                if let Some(neighbor_color) = coloring.get(&neighbor) {
                    if node_color == *neighbor_color {
                        return None
                    }
                } else {
                    queue.push_back(neighbor);
                    coloring.insert(neighbor, node_color.opposite());
                    component.insert(neighbor);
                }
            }
        }
    }

    Some(component)
}

fn is_bipartite(graph: &Graph) -> bool {
    let mut unvisited_nodes = (1..=graph.nr_nodes).collect::<HashSet<_>>();
    while let Some(start_node) = unvisited_nodes.iter().next().cloned() {
        let Some(bipartite_component) = get_consistently_colored_component(graph, start_node) else {
            return false;
        };
        unvisited_nodes.retain(|x| !bipartite_component.contains(x));
    }

    true
}

fn main() {
    let (nr_nodes, nr_edges) = read_pair();
    let edge_list = read_edges(nr_edges as usize);
    let graph = Graph::from_edge_list(nr_nodes, &edge_list);
    let result = is_bipartite(&graph);
    println!("{}", result as i32);
}
